packages/flutter/lib/src/widgets/text_selection.dart

```diff
@@ -813,6 +813,276 @@ class _TextSelectionHandleOverlayState
+  /// force press selection.
+  bool get forcePressEnabled;
+
+  /// Whether selection is enabled.
+  ///
+  /// If false, [TextSelectionGestureDetectorBuilder] will disable certain
+  /// selection handlers
+  bool get selectionEnabled;
+}
+
+/// Builds a [TextSelectionGestureDetector] to wrap an [EditableText].
+///
+/// The class implements sensible defaults for many user interactions
+/// with an [EditableText] (see the documentation of the various gesture handler
+/// methods, e.g. [onTapDown], [onFrocePress], etc.). Subclasses of
+/// [EditableTextSelectionHandlesProvider] can change the behavior performed in
+/// responds to these gesture events by overriding the corresponding handler
+/// methods of this class.
+///
+/// The resulting [TextSelectionGestureDetector] to wrap an [EditableText] is
+/// obtained by calling [buildGestureDetector].
+///
+/// See also:
+///
+///  * [TextField], which uses a subclass to implement the Material-specific
+///    gesture logic of an [EditableText].
+///
+///  * [CupertinoTextField], which uses a subclass to implement the
+///    Cupertino-specific gesture logic of an [EditableText].
+class TextSelectionGestureDetectorBuilder {
+  /// Create a [TextSelectionGestureDetectorBuilder] that provide selection
+  /// handlers to input client.
+  TextSelectionGestureDetectorBuilder({
+    @required this.client,
+  }) : assert(client != null);
+
+  /// The client to be provided a handler for.
+  ///
+  /// Must not be null
+  @protected
+  final TextSelectionGestureDetectorBuilderClient client;
+
+  /// Whether to show the selection tool bar.
+  bool get shouldShowSelectionToolbar => _shouldShowSelectionToolbar;
+  bool _shouldShowSelectionToolbar = true;
+
+  /// The [State] of underlying [EditableText].
+  @protected
+  EditableTextState get editableText => client.editableTextKey.currentState;
+
+  /// The underlying [RenderEditable].
+  @protected
+  RenderEditable get renderEditable => editableText.renderEditable;
+
+  /// Notifies [renderEditable] a tap down is detected and determines whether to
+  /// show toolbar based on signal source.
+  ///
+  /// See also:
+  ///
+  ///  * [TextSelectionGestureDetector.onTapDown], which triggers this callback.
+  @protected
+  void onTapDown(TapDownDetails details) {
+    renderEditable.handleTapDown(details);
+    // The selection overlay should only be shown when the user is interacting
+    // through a touch screen (via either a finger or a stylus). A mouse shouldn't
+    // trigger the selection overlay.
+    // For backwards-compatibility, we treat a null kind the same as touch.
+    final PointerDeviceKind kind = details.kind;
+    _shouldShowSelectionToolbar = kind == null
+                              || kind == PointerDeviceKind.touch
+                              || kind == PointerDeviceKind.stylus;
+  }
+
+  /// Selects words in the range specified in [details] if selection is enabled.
```

Comment:
/// Handler for [TextSelectionGestureDetector.onForcePressStart].
///
/// By default, it selects the word at the position of the force press, if selection is enabled.